### PR TITLE
feat :  Enhancements to MKDocs Rendering

### PR DESCRIPTION
**Pull request by Rakesh-Seenu**
# For authors
image
image

leaderboard website for reference: https://rakesh-seenu.github.io/VPELeaderboard/

## Description

# Pull Request: Enhancements to MKDocs Rendering, Data Processing, and Algorithm Table Integration Summary of Changes & Updates in This PR

This PR introduces several improvements and feature updates to the MKDocs rendering workflow, data processing pipeline, and algorithm leaderboard table. The key changes implemented are :

1. MKDocs HTML Rendering Integration
- Implemented a structured flow where Python scripts generate output, which is then converted into Markdown format for further processing by MKDocs.
- This ensures that documentation updates are dynamically rendered without manual intervention.

2. Refactored Project Structure
- Introduced two new folders, algorithms and data, under the vpeleaderboard directory to better organize Python files.
Automated XML Data Extraction & Tabular Representation

3. Developed a mechanism to scan and process XML files within a designated folder.
- Extracted key metadata such as abstract, number of species, number of parameters, and model name.
- These extracted values are structured into a table and integrated into the data section of the leaderboard for enhanced visibility and accessibility.

4. Dynamic Algorithm Table Generation
- Previously, the algorithm leaderboard relied on a hardcoded Markdown file.
- Now, the table is dynamically generated by processing mock data, ensuring that updates to the dataset are reflected in real time on the frontend.

5. Expanded Documentation for Data Processing Scripts
- Added comprehensive documentation within the data section for two core Python files:
  - basico_model.py
  - sys_bio_model.py
- This documentation improves clarity on how these models function and interact with the system.

# Planned Updates for the Next PR :
1. Refining the Abstract Table
Further processing and enhancement are required to improve the structure and formatting of the abstract table for better readability.

2. Implementing Configuration & Hydra Files
Planning to replace hardcoded file paths within the code by introducing configuration files and Hydra-based configurations.
This will enhance flexibility and maintainability, allowing for environment-specific configurations without code modifications.

3. Developing a Data Loader Module
Implementing a dedicated data loader to streamline the ingestion and preprocessing of datasets.
This will standardize data access across different components of the project.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [*] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [*] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [*] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [*] My code follows the style guidelines mentioned in the Code/DevOps guides
- [*] I have commented my code, particularly in hard-to-understand areas
- [*] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [*] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [*] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Comment by gurdeep330** [comment]
@Rakesh-Seenu 
image


The code coverage is not 100%. Do you mind fixing this?

**Comment by Rakesh-Seenu** [comment]
> @Rakesh-Seenu image
> 
> The code coverage is not 100%. Do you mind fixing this? Thx!
> 
> Also, I just noticed that the header of the table under the Data tab is "Algorithm Metric Table".

Hi @gurdeep330, I have made the changes requested and corrected the tests coverage.

File: tests/test_markdown.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+A test markdown class for pytest unit testing.
+"""
+
+import os
+import pytest
+from vpeleaderboard.data.src.basico_model import BasicoModel
+from app.data.markdown_report import generate_markdown_report
+
+@pytest.fixture(name="model")
+def model_fixture():
```
Comment (dmccloskey) [inline tests/test_markdown.py:11]: Would it be possible to consolidate the common fixture/mock functions/classes into a single utils (or similar) module to reduce code repitition?

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Some solid enhancements indeed 👍. I just had a few minor comments.

File: app/data/markdown_report.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    biomodel = BasicoModel(sbml_folder_path=folder_path)
+    metadata_list = [
+        biomodel.get_model_metadata(file)
+        for file in os.listdir(folder_path)
+        if file.endswith(".xml")
+        ]
+    metadata_df = pd.DataFrame(metadata_list)
+    metadata_records = metadata_df.to_dict(orient='records')
+    markdown_content = create_markdown(metadata_records, "app/templates", "data.txt")
+    print("DEBUG: Rendered HTML ->", markdown_content)
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/data/markdown_report.py:38]: no, we don't need the print statement.

File: app/overrides/main.html [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  function startIntro(){
+      introJs().setOptions({
+          tooltipClass: 'customTooltip'
+      }).start();
+  }
+</script>
+<script>
+  
+
+  // new DataTable('#table1', {
```
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/overrides/main.html:58]: No, I will remove them.

File: app/templates/algo.txt
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/templates/algo.txt:1]: yes, I have changed it to html extension.

File: tests/test_markdown.py (hunk 1 above)
Comment (Rakesh-Seenu) [reply tests/test_markdown.py:11]: Hi @dmccloskey, I have created a new utils.py file to consolidate all the fixtures. Could you please review and let me know if everything looks correct?

File: app/templates/data.txt
[the remark this answers is not in the document]
Comment (Rakesh-Seenu) [reply app/templates/data.txt:1]: yes, I have changed it to html extension.
